From: Filipe Manana <fdmanana@suse.com>
Date: Tue, 12 Jul 2016 14:45:02 +0100
Patch-mainline: Not yet, used for debugging issues in openqa only
References: bsc#985850
Subject: [PATCH] Btrfs: send, add some error messages before hitting BUG_ONs

Just to give some information about what went wrong and allow further
debugging of the issue.

Signed-off-by: Filipe Manana <fdmanana@suse.com>
---
 fs/btrfs/send.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/fs/btrfs/send.c b/fs/btrfs/send.c
index 63a6152..482ede2 100644
--- a/fs/btrfs/send.c
+++ b/fs/btrfs/send.c
@@ -5601,6 +5601,15 @@ static int changed_ref(struct send_ctx *sctx,
 {
 	int ret = 0;
 
+	if (sctx->cur_ino != sctx->cmp_key->objectid)
+		btrfs_err(sctx->send_root->fs_info,
+			  "unexpected ref for different inode, current inode is %llu, found inode is %llu, send root %llu, parent root %llu, result %d",
+			  sctx->cur_ino, sctx->cmp_key->objectid,
+			  sctx->send_root->root_key.objectid,
+			  sctx->parent_root ?
+			  sctx->parent_root->root_key.objectid : 0,
+			  result);
+
 	BUG_ON(sctx->cur_ino != sctx->cmp_key->objectid);
 
 	if (!sctx->cur_inode_new_gen &&
@@ -5626,6 +5635,15 @@ static int changed_xattr(struct send_ctx *sctx,
 {
 	int ret = 0;
 
+	if (sctx->cur_ino != sctx->cmp_key->objectid)
+		btrfs_err(sctx->send_root->fs_info,
+			  "unexpected xattr for different inode, current inode is %llu, found inode is %llu, send root %llu, parent root %llu, result %d",
+			  sctx->cur_ino, sctx->cmp_key->objectid,
+			  sctx->send_root->root_key.objectid,
+			  sctx->parent_root ?
+			  sctx->parent_root->root_key.objectid : 0,
+			  result);
+
 	BUG_ON(sctx->cur_ino != sctx->cmp_key->objectid);
 
 	if (!sctx->cur_inode_new_gen && !sctx->cur_inode_deleted) {
@@ -5650,6 +5668,15 @@ static int changed_extent(struct send_ctx *sctx,
 {
 	int ret = 0;
 
+	if (sctx->cur_ino != sctx->cmp_key->objectid)
+		btrfs_err(sctx->send_root->fs_info,
+			  "unexpected extent for different inode, current inode is %llu, found inode is %llu, send root %llu, parent root %llu, result %d",
+			  sctx->cur_ino, sctx->cmp_key->objectid,
+			  sctx->send_root->root_key.objectid,
+			  sctx->parent_root ?
+			  sctx->parent_root->root_key.objectid : 0,
+			  result);
+
 	BUG_ON(sctx->cur_ino != sctx->cmp_key->objectid);
 
 	if (!sctx->cur_inode_new_gen && !sctx->cur_inode_deleted) {
-- 
1.8.4.5

